network/pubsubchannel/pubsub.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2019 The Swarm Authors
+// This file is part of the Swarm library.
+//
+// The Swarm library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Swarm library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Swarm library. If not, see <http://www.gnu.org/licenses/>.
+package pubsubchannel
+
+import (
+	"strconv"
+	"sync"
+
+	"github.com/ethersphere/swarm/log"
+)
+
+//PubSubChannel represents a pubsub system where subscriber can .Subscribe() and publishers can .Publish() or .Close().
+type PubSubChannel struct {
+	subscriptions []*Subscription
+	subsMutex     sync.RWMutex
+	nextId        int
+	quitC         chan struct{}
+}
+
+// Subscription is created in PubSubChannel using pubSub.Subscribe(). Subscribers can receive using .ReceiveChannel().
+// or .Unsubscribe()
+type Subscription struct {
+	closed    bool
+	pubSubC   *PubSubChannel
+	signal    chan interface{}
+	closeOnce sync.Once
+	id        string
+	lock      sync.RWMutex
+}
+
+// New creates a new PubSubChannel.
+func New() *PubSubChannel {
+	return &PubSubChannel{
+		subscriptions: make([]*Subscription, 0),
+		quitC:         make(chan struct{}, 1),
```

Comment:
Why a this channel has a buffer? Shouldn't it be just closed somewhere?

Reply:
My fault. Let me correct this.